### PR TITLE
New module for managing printers in CUPS

### PR DESCRIPTION
**Pull request by multimac**
Creates a module for installing and uninstalling printers in CUPS. Can handle most of the functionality in lpadmin.

Has been tested on CUPS 1.7.2 and 2.0

### CODE REVIEW

**Comment by gregdek** [comment]
@Multimac Thanks for submitting this new module to Ansible Extras! This module is now in community review, a process that is open to all Ansible users. In order for this module to be approved, it must gain the following votes:

“works_for_me”: If you have tested the module thoroughly, including testing of all of the module’s options, and if the module works for you, please add “works_for_me” in the comments.

“passes_guidelines”: If you have gone through the module guidelines and the module meets all of the requirements, please add “passes_guidelines” in the comments. Guidelines are available here: http://docs.ansible.com/developing_modules.html#module-checklist

“needs_revision”: If the module fails to work for you, or if it doesn’t meet guidelines, please add “needs_revision” in the comments with details about what needs to be fixed.

When a module has both “works_for_me” and “passes_guidelines” tags, we will promote the module for inclusion in Ansible Extras.  At this point, you will be expected to maintain the module by fixing bugs and evaluating pull requests in a timely manner. 

Thanks again for submitting your Ansible module!

File: system/cups_lpadmin.py [hunk 1]
```diff
@@ -0,0 +1,393 @@
...
+import json
+import re
+import shlex
+
+DOCUMENTATION = '''
+---
+module: cups_lpadmin
+author: David Symons (Multimac) <Mult1m4c@gmail.com>
+short_description: Manages printers in CUPS via lpadmin
+version_added: "1.9.2"
```
Comment (resmo) [inline system/cups_lpadmin.py]: should be `version_added: "2.0"`

**Comment by k0ste** [comment]
works_for_me
Tested on cups 2.1.3.

Thanks for module.
I think the 'dest' should be named 'name'...
And may be good idea (or not) is add 'proto' option, because I see in code - now if not ':\\\' then proto is lpd. How it *may* look:

Connect remote printer via ipp (uri is **ipp://192.168.2.60:631/printers/HP_P2055**):
```yaml
cups_lpadmin:
  state: 'present'
  name: 'HP_P2055'
  model: 'raw'
  proto: 'ipp'
  uri: '192.168.1.1'
```
```python
uri = proto + '://' + uri + '/printers/' + name
```
Is good for ipp, because use default port. Bot not good if used port 631 or if proto is socket with port 9100, just idea, it's more easy for definition, and may be defined via jinja.

**Comment by k0ste** [comment]
Well, I tested module for change options, states.
Add 'default' param for default server printer definition. Rename 'dest' to 'name' and makeup doc strings. In uri I remove one '/', because if the uri like 'hp:/usb' this is fails. As I wrote before, not sure about this in uri is good idea.
From this day, this module in production on my work, thanks again @Multimac.
Patch included:

```diff
diff --git a/my_modules/cups_lpadmin.py b/my_modules/cups_lpadmin.py
index 730ae04..f0ddccd 100644
--- a/my_modules/cups_lpadmin.py
+++ b/my_modules/cups_lpadmin.py
@@ -18,10 +18,6 @@
 # You should have received a copy of the GNU General Public License
 # along with this software.  If not, see <http://www.gnu.org/licenses/>.
 
-import json
-import re
-import shlex
-
 DOCUMENTATION = '''
 ---
 module: cups_lpadmin
@@ -31,74 +27,96 @@ version_added: "2.0"
 requirements:
   - CUPS 1.7+
 description:
-  - Creates, removes and sets options for printers in CUPS
+    - Creates, removes and sets options for printers in CUPS.
 options:
-  state:
-    choices: [present, absent]
-    default: present
-    description:
-      - Whether the printer should or should not be in CUPS.
-  dest:
-    description:
-      - The destination to configure or remove from CUPS
-    required: true
-  uri:
-    default: None
-    description:
-      - The URI to use when connecting to the printer
-      - This is only required in the present state
-    required: false
-  enabled:
-    default: true
-    description:
-      - Whether or not the printer should be enabled and accepting jobs
-    required: false
-  shared:
-    default: false
-    description:
-      - Whether or not the printer should be shared on the network
-    required: false
-  model:
-    default: None
-    description:
-      - The System V interface or PPD file to be used for the printer
-    required: false
-  info:
-    default: None
-    description:
-      - The textual description of the printer.
-    required: false
-  location:
-    default: None
-    description:
-      - The textual location of the printer.
-    required: false
-  options:
-    default: { }
-    description:
-      - A dictionary of key-value pairs describing printer options and their required value.
-    required: false
+    name:
+        description:
+            - Name of the printer in CUPS
+        required: true
+        default: null
+    state:
+        description:
+            - Whether the printer should or not be in CUPS.
+        required: false
+        default: present
+        choices: ["present", "absent"]
+    uri:
+        description:
+            - The URI to use when connecting to the printer. This is only required in the present state.
+        required: false
+        default: null
+    enabled:
+        description:
+            - Whether or not the printer should be enabled and accepting jobs.
+        required: false
+        default: true
+        choices: ["true", "false"]
+    shared:
+        description:
+            - Whether or not the printer should be shared on the network.
+        required: false
+        default: false
+        choices: ["true", "false"]
+    model:
+        description:
+            - The System V interface or PPD file to be used for the printer.
+        required: false
+        default: null
+    default:
+        description:
+          - Set default server printer. Only one printer can be default.
+        required: false
+        default: false
+        choices: ["true", "false"]
+    info:
+        description:
+            - The textual description of the printer.
+        required: false
+        default: null
+    location:
+        description:
+            - The textual location of the printer.
+        required: false
+        default: null
+    options:
+        description:
+            - A dictionary of key-value pairs describing printer options and their required value.
+        default: { }
+        required: false
 '''
 
 EXAMPLES = '''
+# Creates HP printer via ethernet, set default paper size and
+  make this printer as server default
+- cups_lpadmin:
+    name: 'HP_M1536'
+    state: 'present'
+    uri: 'hp:/net/HP_LaserJet_M1536dnf_MFP?ip=192.168.1.2'
+    model: 'drv:///hp/hpcups.drv/hp-laserjet_m1539dnf_mfp-pcl3.ppd'
+    default: 'true'
+    location: 'Lib'
+    info: 'MFP'
+    options:
+      media: 'iso_a4_210x297mm'
+
 # Creates a Zebra ZPL printer called zebra
-- cups_lpadmin: state=present dest=zebra uri=192.168.1.2 model=drv:///sample.drv/zebra.ppd
+- cups_lpadmin: state=present name=zebra uri=192.168.1.2 model=drv:///sample.drv/zebra.ppd
 
 # Updates the zebra printer with some custom options
 - cups_lpadmin:
     state: present
-    dest: zebra
+    name: zebra
     uri: 192.168.1.2
     model: drv:///sample.drv/zebra.ppd
     options:
       PageSize: w288h432
 
 # Creates a raw printer called raw_test
-- cups_lpadmin: state=present dest=raw_test uri=192.168.1.3
+- cups_lpadmin: state=present name=raw_test uri=192.168.1.3
 
 # Deletes the printers set up by the previous tasks
-- cups_lpadmin: state=absent dest=zebra
-- cups_lpadmin: state=absent dest=raw_test
+- cups_lpadmin: state=absent name=zebra
+- cups_lpadmin: state=absent name=raw_test
 '''
 
 class CUPSPrinter(object):
@@ -106,11 +124,12 @@ class CUPSPrinter(object):
     def __init__(self, module):
         self.module = module
 
-        self.destination = module.params['dest']
+        self.name = module.params['name']
         self.uri = module.params['uri']
 
         self.enabled = module.params['enabled']
         self.shared = module.params['shared']
+        self.default = module.params['default']
 
         self.model = module.params['model']
 
@@ -120,7 +139,7 @@ class CUPSPrinter(object):
         self.options = module.params['options']
 
         # Use lpd if a protocol is not specified
-        if self.uri and '://' not in self.uri:
+        if self.uri and ':/' not in self.uri:
             self.uri = 'lpd://{0}/'.format(self.uri)
 
     def _get_installed_drivers(self):
@@ -159,8 +178,8 @@ class CUPSPrinter(object):
         return drivers
 
     def _get_make_and_model(self):
-        if not self.model:
-            # We're dealing with a raw printer then
+        '''Raw printer is defined or model not defined'''
+        if not self.model or self.model == 'raw':
             return "Local Raw Printer"
 
         installed_drivers = self._get_installed_drivers()
@@ -170,9 +189,7 @@ class CUPSPrinter(object):
         self.module.fail_json(msg="unable to determine printer make and model")
 
     def _install_printer(self):
-        cmd = [ 'lpadmin',
-                '-p', self.destination,
-                '-v', self.uri ]
+        cmd = ['lpadmin', '-p', self.name, '-v', self.uri]
 
         if self.enabled:
             cmd.append('-E')
@@ -184,30 +201,34 @@ class CUPSPrinter(object):
 
         if self.model:
             cmd.extend(['-m', self.model])
-
         if self.info:
             cmd.extend(['-D', self.info])
         if self.location:
             cmd.extend(['-L', self.location])
+        if self.default:
+            cmd.extend(['-d', self.name])
 
         return self.module.run_command(cmd)
 
     def _install_printer_options(self):
-        cmd = [ 'lpadmin',
-                '-p', self.destination ]
+        cmd = ['lpadmin', '-p', self.name]
 
         for k, v in self.options.iteritems():
             cmd.extend(['-o', '{0}={1}'.format(k, v)])
 
+        '''Target printer is default server printer'''
+        if self.default:
+            cmd.extend(['-d', self.name])
+
         return self.module.run_command(cmd)
 
     def _uninstall_printer(self):
-        cmd = ['lpadmin', '-x', self.destination]
+        cmd = ['lpadmin', '-x', self.name]
         return self.module.run_command(cmd)
 
     def get_printer_cups_options(self):
-        """Returns the CUPS options for the printer"""
-        cmd = ['lpoptions', '-p', self.destination]
+        '''Returns the CUPS options for the printer'''
+        cmd = ['lpoptions', '-p', self.name]
         (rc, out, err) = self.module.run_command(cmd)
 
         options = { }
@@ -222,8 +243,8 @@ class CUPSPrinter(object):
         return options
 
     def get_printer_specific_options(self):
-        """Returns the printer specific options for the printer, as well as the accepted options"""
-        cmd = ['lpoptions', '-p', self.destination, '-l']
+        '''Returns the printer specific options for the printer, as well as the accepted options'''
+        cmd = ['lpoptions', '-p', self.name, '-l']
         (rc, out, err) = self.module.run_command(cmd)
 
         options = { }
@@ -262,11 +283,11 @@ class CUPSPrinter(object):
             'printer-location': self.location,
         }
 
-        # 'printer-info' defaults to the destination name if not specified manually
+        # 'printer-info' defaults to the printer name if not specified manually
         if self.info:
             expected_cups_options['printer-info'] = self.info
         else:
-            expected_cups_options['printer-info'] = self.destination
+            expected_cups_options['printer-info'] = self.name
 
         if self.shared:
             expected_cups_options['printer-is-shared'] = 'true'
@@ -297,10 +318,10 @@ class CUPSPrinter(object):
         return True
 
     def exists(self):
-        cmd = ['lpstat', '-p', self.destination]
+        cmd = ['lpstat', '-p', self.name]
         (rc, out, err) = self.module.run_command(cmd)
 
-        # This command will fail if the destination doesn't exist (rc != 0)
+        # This command will fail if the self.name doesn't exist (rc != 0)
         return rc == 0
 
     def install(self):
@@ -346,10 +367,11 @@ def main():
     module = AnsibleModule(
         argument_spec = dict(
             state=dict(default='present', choices=['present', 'absent'], type='str'),
-            dest=dict(required=True, type='str'),
+            name=dict(required=True, type='str'),
             uri=dict(default=None, type='str'),
             enabled=dict(default=True, type='bool'),
             shared=dict(default=False, type='bool'),
+            default=dict(default=False, type='bool'),
             model=dict(default=None, type='str'),
             info=dict(default=None, type='str'),
             location=dict(default=None, type='str'),
@@ -366,7 +388,7 @@ def main():
 
     result = { }
     result['state'] = module.params['state']
-    result['destination'] = cups_printer.destination
+    result['name'] = cups_printer.name
 
     state = module.params['state']
     if state == 'present':
```

**Comment by sistemasdistal** [comment]
needs_revision

Tested on ubuntu 14.04, the module fails if the output for `lpinfo -l -m` is localized.

Working around it with an `environment:` section in the task allows the module to be used:

```
- name: add printer
  environment:
    LC_ALL: en_US
  cups_lpadmin: state=present dest="testprinter" uri="socket://printerhost" model="driver.ppd"

```

**Comment by sistemasdistal** [comment]
needs_revision

Always reports changed state when using `location`... apparently because my lpoptions doesn't print the printer-location:

```
# lpoptions -p q_laser -o printer-location=test
# lpoptions -p q_laser
copies=1 device-uri=socket://q_laser finishings=3 job-hold-until=no-hold job-priority=50 job-sheets=none,none marker-change-time=0 number-up=1 ppd-timestamp=* printer-commands=AutoConfigure,Clean,PrintSelfTestPage printer-info=q_laser printer-is-accepting-jobs=true printer-is-colormanaged=true printer-make-and-model='Generic PCL 3 Printer Foomatic/pcl3 (recommended)' printer-state=3 printer-state-change-time=1459531191 printer-state-reasons=none printer-type=10530844 printer-uri-supported=ipp://localhost:631/printers/q_laser

```

My cups version is 1.7.2-0ubuntu1.7

Will be looking into it and get back with a patch.

**Comment by HP41** [comment]
I was wondering when the merge will happen/when can we natively control printers via ansible?
Thanks!

**Comment by k0ste** [comment]
works_for_me

I fork this module and add extra features - f926233b7c32df19e5f8ce24fdbc5431b23e45e7.
If somebody else have interest to use this in production (we use).

**Comment by HP41** [comment]
works_for_me

However I've forked it from @k0ste above and added some more features such as support for classes - https://github.com/HP41/ansible-modules-extras/commit/df96f9273c6a8b5e3fda60ca6604a1d603adcd1c
Thanks guys!

**Comment by dustymabe** [comment]
Hey @multimac can you incorporate any useful additions by @k0ste and @HP41 and update this PR?

**Comment by mathieumd** [comment]
needs_revision

I tried HP41@df96f92 version:
- it works the first time;
- for next times:
  - I had to modify line 383 (BEFORE `kv = map(str.strip, kv)` / AFTER `kv = list(map(str.strip, kv))`) or it would fail with the error below[1].
  - And just like @sistemasdistal, it reports change status every times.
1. The error:

``` json
{
  "failed": true,
  "item": {
    "default": "true",
    "info": "Toshiba e-Studio 230",
    "location": "Couloir du 1er",
    "model": "foomatic-db-compressed-ppds:0/ppd/foomatic-ppd/Toshiba-e-Studio_230-Postscript.ppd",
    "name": "Noir_Blanc_1",
    "uri": "192.168.1.51"
  },
  "module_stderr": "",
  "module_stdout": "Traceback (most recentcall last):
  File \"/tmp/ansible__p6OPz/ansible_module_cups_lpadmin.py\", line 990, in <module>
    main()
  File \"/tmp/ansible__p6OPz/ansible_module_cups_lpadmin.py\", line 971, in main
    (rc, out, err) = cups_object.printer_install()
  File \"/tmp/ansible__p6OPz/ansible_module_cups_lpadmin.py\", line 833, in printer_install
    if self.exists() and not self.printer_check_cups_options():
  File \"/tmp/ansible__p6OPz/ansible_module_cups_lpadmin.py\", line 654, in printer_check_cups_options
    'printer-make-and-model': self._printer_get_make_and_model(),
  File \"/tmp/ansible__p6OPz/ansible_module_cups_lpadmin.py\", line 411, in _printer_get_make_and_model
    installed_drivers = self._printer_get_installed_drivers()
  File \"/tmp/ansible__p6OPz/ansible_module_cups_lpadmin.py\", line 385, in _printer_get_installed_drivers
    curr[kv[0]] = kv[1]
TypeError: 'itertools.imap' object has no attribute ''
",
  "msg": "MODULE FAILURE"
}
```

**Comment by k0ste** [comment]
@mathieumd, try last revision of module https://github.com/k0ste/ansible-modules-extras/commit/35473942e2f6a60476c1bb5af502783f5a077e55.

**Comment by mathieumd** [comment]
@k0ste I finally tried the revision you linked (sorry for the delay): while the bug at line 383 appears to be solved, it still reports change status at every runs.

**Comment by k0ste** [comment]
@mathieumd, you make printer as server default. Is no way in cups detect current value of this key. For this reason, the action will always be executed.

**Comment by mathieumd** [comment]
Actually, you can use `lpstat -d` to get the current default printer. ;-)

**Comment by k0ste** [comment]
> Actually, you can use `lpstat -d` to get the current default printer. ;-)

Suddenly!

**Pull request closed without merging**